lpc55_sign/src/bin/lpc55_sign.rs

```diff
@@ -135,80 +153,82 @@ fn main() -> Result<()> {
+            dice_args,
+            image_args:
+                ImageArgs {
+                    address,
+                    src_bin,
+                    ref dest_bin,
+                    ref dest_cmpa,
+                    ref dest_cfpa,
+                },
+            cert_cfg,
         } => {
-            let cfg_contents = std::fs::read(&cfg)?;
-            let toml: CfgFile = toml::from_slice(&cfg_contents)?;
-
-            let rkth = signed_image::sign_chain(&src_bin, None, &toml.certs, &dest_bin, address)?;
-            signed_image::create_cmpa(
-                with_dice,
-                with_dice_inc_nxp_cfg,
-                with_dice_cust_cfg,
-                with_dice_inc_sec_epoch,
-                &rkth,
-                &dest_cmpa,
-            )?;
-
-            println!(
-                "Done! Signed image written to {:?}, CMPA to {:?}",
-                &dest_bin, &dest_cmpa
-            );
+            let cfg_contents = std::fs::read(&cert_cfg)?;
+            let cfg: CertConfig = toml::from_slice(&cfg_contents)?;
+            let private_key = std::fs::read_to_string(&cfg.private_key)?;
+            let signing_certs = read_certs(&cfg.signing_certs)?;
+            let mut root_certs = read_certs(&cfg.root_certs)?;
+            if root_certs.len() < 4 {
+                let empty_roots = [vec![], vec![], vec![], vec![]];
+                root_certs.extend_from_slice(&empty_roots[..4 - root_certs.len()]);
+            } else if root_certs.len() > 4 {
+                bail!("Too many roots, max four");
+            }
+
+            let image = std::fs::read(src_bin)?;
+            let stamped =
+                signed_image::stamp_image(image, signing_certs, root_certs.clone(), address)?;
+            let signed = signed_image::sign_image(&stamped, &private_key)?;
+            std::fs::write(dest_bin, signed)?;
+            check!(OK, "Signed image written to {}", dest_bin.display());
+
+            if let Some(dest_cmpa) = dest_cmpa {
+                let rotkh = signed_image::root_key_table_hash(root_certs.clone())?;
+                std::fs::write(dest_cmpa, generate_cmpa(dice_args, rotkh)?.to_vec()?)?;
+                println!("CMPA written to {}", dest_cmpa.display());
+            }
+            if let Some(dest_cfpa) = dest_cfpa {
+                std::fs::write(dest_cfpa, generate_cfpa(root_certs)?.to_vec()?)?;
+                println!("CFPA written to {}", dest_cfpa.display());
+            }
         }
         Command::SignedImage {
-            with_dice,
-            with_dice_inc_nxp_cfg,
-            with_dice_cust_cfg,
-            with_dice_inc_sec_epoch,
-            src_bin,
-            priv_key,
-            root_cert0,
-            dest_bin,
-            dest_cmpa,
-            address,
-            cfpa,
+            dice_args,
+            image_args:
+                ImageArgs {
+                    address,
+                    src_bin,
+                    ref dest_bin,
```

Comment:
(random comment: I'm surprised to see `ref` patterns here, they're pretty rare in Rust nowadays -- do you need them?)

Reply:
No, I don't know why I reached for them over a few `&`s! Removed in b794e44, thanks.